modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/Cluster/ClientClusterGroup.cs

```diff
@@ -0,0 +1,311 @@
+        /** Node ids collection. */
+        private IList<Guid> _nodeIds;
+
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="ignite">Ignite.</param>
+        /// <param name="marsh">Marshaller.</param>
+        internal ClientClusterGroup(IgniteClient ignite, Marshaller marsh)
+            : this(ignite, marsh, ClientClusterGroupProjection.Empty)
+        {
+        }
+
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="ignite">Ignite.</param>
+        /// <param name="marsh">Marshaller.</param>
+        /// <param name="projection">Projection.</param>
+        /// <param name="predicate">Predicate.</param>
+        private ClientClusterGroup(IgniteClient ignite, Marshaller marsh,
+            ClientClusterGroupProjection projection, Func<IClusterNode, bool> predicate = null)
+        {
+            Debug.Assert(ignite != null);
+            Debug.Assert(marsh != null);
+
+            _ignite = ignite;
+            _marsh = marsh;
+            _projection = projection;
+            _predicate = predicate;
+        }
+
+        /** <inheritDoc /> */
+        public IClientClusterGroup ForAttribute(string name, string val)
+        {
+            IgniteArgumentCheck.NotNullOrEmpty(name, "name");
+
+            return new ClientClusterGroup(_ignite, _marsh, _projection.ForAttribute(name, val));
+        }
+
+        /** <inheritDoc /> */
+        public IClientClusterGroup ForDotNet()
+        {
+            return ForAttribute(AttrPlatform, Platform);
+        }
+
+        /** <inheritDoc /> */
+        public IClientClusterGroup ForServers()
+        {
+            return new ClientClusterGroup(_ignite, _marsh, _projection.ForServerNodes());
+        }
+
+        /** <inheritDoc /> */
+        public IClientClusterGroup ForPredicate(Func<IClusterNode, bool> p)
+        {
+            IgniteArgumentCheck.NotNull(p, "p");
+
+            var newPredicate = _predicate == null ? p : node => _predicate(node) && p(node);
+            return new ClientClusterGroup(_ignite, _marsh, _projection, newPredicate);
+        }
+
+        /** <inheritDoc /> */
+        public ICollection<IClusterNode> GetNodes()
+        {
+            return RefreshNodes();
+        }
+
+        /** <inheritDoc /> */
+        public IClusterNode GetNode(Guid id)
+        {
+            if (id == Guid.Empty)
+            {
+                throw new ArgumentException("Node id should not be empty.");
+            }
+
+            return GetNodes().FirstOrDefault(node => node.Id == id);
+        }
+
+        /** <inheritDoc /> */
+        public IClusterNode GetNode()
+        {
+            return GetNodes().FirstOrDefault();
+        }
+
+        /// <summary>
+        /// Refresh projection nodes.
+        /// </summary>
+        /// <returns>Nodes.</returns>
+        private IList<IClusterNode> RefreshNodes()
+        {
+            long oldTopVer = Interlocked.Read(ref _topVer);
```

Comment:
Let's simplify things and get rid of topVer handling. Just request new ids, that's it.